### DISCUSSION TITLE
Release process

### DISCUSSION THREAD

**Discussion by pepicrft** [MobileNativeFoundation/discussions#67] [General]
Hi there!
I haven't seen any discussion around this topic so I thought it'd be a good idea to start one to see what other companies are doing.

I work at @shopify, and a few years ago we decided to move away from CI pipelines as the place to codify the release process. We have several teams doing mobile development, and each had its own release logic, which in some cases was complex and brittle. When things failed, they required support from my team, mobile tooling, to unlock their release.

We set out to build an internal solution for coordinating releases. We called it Shipit Mobile. The goal was to standardize the process across all the projects and own as much as possible from it to make sure that we can optimize it and improve it without much work on the product team's side. The natural technology choice for us was a Rails app. We wanted it to be server-side to be able to build integrations with Buildkite, GitHub, Slack, and Google Cloud. 

This is the **release process** that we proposed:

- Developers go to their project on the platform.
- From the list of commits, they pick the one that they'd like to release.
- Shipit Mobile creates a release branch and a commit with the version bump. The resulting build from that commit is called "candidate"
- A release might require several candidates until there's one that is uploaded to Google Play / App Store Connect.
- The process of triggering candidates is as easy as pushing commits to the release branch. Shipit Mobile receives webhooks from GitHub and turns them into a build of the app bundle.
- Builds run as CI pipelines on Buildkite. We subscribe to Buildkite's webhooks to get updates about the status of the build.
- When a candidate is built, Shipit Mobile takes care of uploading the assets to App Store Connect and Google Play, as well as the dSYMs and source maps to the error tracking platform (in our case Bugsnag).
- The developer takes the final steps on the developer portal. Once the app is released, they mark it as complete on Shipit Mobile and we create and commit and tag it with the version that has been released.

Because we own and design the service, we could design a UI that is friendly for people that are not developers. Moreover, we could build integrations with Slack. For example, we notify a developer when a candidate is ready to be published, or when Apple has finished processing the artifact.

Later on, we decided to move away from App Center, and leverage Shipit Mobile for doing internal releases too. We called them **snapshot builds**. Unlike the traditional "nightly builds" that are triggered from the main branch, we allow developers to trigger builds from any commit. They do it from the UI of Shipit Mobile, but we have plans to support triggering from PRs. Because we are not using App Center, we made it super easy to install a build. Basically, you only have to sign in with your Google account and that's it. Moreover, we are building features on top of it like the possibility of providing feedback after installing a snapshot build.

One thing worth mentioning in regards to how we do internal builds is that we made the build number an implementation detail. We noticed that using the build number to refer to a build adds unnecessary friction:
- Oh, so you found a bug? What's the build number of your build?
- Where do I find the build number?
- Oh, let me find what commit/branch that build number is tied to.

Instead, when Shipit Mobile builds those builds, we pass contextual information such as the commit, branch, PR, author... That way you can easily know what you have on your phone. 

In hindsight, building this service was a great idea because we have several teams doing app development, but it's a fairly large undertaking if you are a small company.

**How are you doing store and internal releases?**

**Comment by mattrobmattrob** [comment]
I think Strava's mobile release process is a well oiled machine at this point. We've been doing it this way for years. I'll talk through a couple different classes of builds (from the iOS context, but Android is on the same schedule).

- String translations aren't really mentioned much below but they're essentially always being translated and always being updated in both the current release and the `main` branch. It's rare that we have mobile engineers actually thinking about strings and the translation process.
- Crashes aren't really mentioned much below but we upload them to our chosen crash service for all builds. We tag them differently so we can filter our crashes by build type and versions. We have high level of priority to crashes that occur in the alpha build since, in theory, it's early signal of issues especially those that are considered "new."

**Alpha (employees):**

Automated build runs at 2 AM PT every day (including weekends), this can be kicked off by any employee manually (mostly done by mobile engineers and rarely) through a Slack command that talks to our in-house build orchestrator/scheduler. This will create a new build on a macOS host, uploaded the dSYMs to our crash tracker, and push the build to an S3 bucket using `fastlane`'s [`aws_s3` plugin](https://github.com/fastlane-community/fastlane-plugin-s3). The alpha iOS app build has special code that will check the latest version in the `version.json` file uploaded to S3 as part of the `aws_s3` plugin. When a newer version is available (since we manually provision all employee devices at this point) and the employee confirms the update, it will redirect the employee to the `itms-services://?action=download-manifest&url=<foo>` URL that's in the `version.json` payload to update the application.

**Adhoc (employees):**

These are essentially the same as alpha except they're delivered via a separate S3 bucket and triggered manually by engineers when they want coworkers to test things. Largely this is done for pre-merge checks of designs or similar functionality. The engineer will have a URL available to send to anyone that's provisioned for the alpha builds that they can use to install this on their device(s).

**Beta (external user, press, and manual testing services):**

We utilize an "early" TestFlight submission to increase TestFlight beta approval speed when it's the real deal. We update it majorly once a week with a smaller update on Thursday for things like translations. This is also the build our external testing services runs so it's important that this is treated as a release-candidate. Similarly to the alpha, this build and submission to iTunes Connect is all scheduled and requires little hands-on maintenance from the person that's in charge of the release for the week.

This build is shipped to all our external beta users through TestFlight. We utilize Zendesk and the rarely-checked-build-in-TestFlight-feedback to gather real world user feedback on the build.

|Week|M|T|W|T|F|S|S|
|---|---|---|---|---|---|---|---|
|1|Build and ship "fake" beta to TestFlight with version (`N+1`) that'll be next.||Create release branch. Build new "real" beta (`N+1`) and submit to TestFlight for external release once it's processed.|Build new "real" beta (`N+1`) and submit to TestFlight for external release once it's processed.||||
|2|Build and ship "fake" beta to TestFlight with version (`N+2`) that'll be next.||Create release branch. Build new "real" beta (`N+2`) and submit to TestFlight for external release once it's processed.|Build new "real" beta (`N+2`) and submit to TestFlight for external release once it's processed.||||

**Release (all users):**
This is similar to the beta except it's shipped most optimized and built as minimally as possible. All actions are automatically ran on a  schedule except for the "Release" button in iTunes Connect, which can be ran from a Slack command by the person in charge of the release.

|Week|M|T|W|T|F|S|S|
|---|---|---|---|---|---|---|---|
|1|||Create release branch (`N`). Release `N-1` with a Slack command based on the metrics for the past week of testing.||||Build and submit `N` for review to iTunes Connect.||
|2|||Create release branch (`N+1`). Release `N` with a Slack command based on the metrics for the past week of testing.||||Build and submit `N+1` for review to iTunes Connect.||

**Comment by JoeSSS** [comment]
hey @pepibumur , there is a lot of nice info about Shipit mobile, but as far as I can see it is not open-sourced, are there any plans to open source it? I believe that it is built on top of the shipit-core that you have in OSS? How much is on top of the core to make it work for the mobile use case? Would you recommend somebody to build it

**Reply by pepicrft** [reply]
Hey Joe!
We couldn't open-source because there is some dependency on internal infrastructure and tools. I don't think open-sourcing will happen because we are designing it for Shopify's needs and processes. If we change our opinion I can let you know.

It's not built on top of the Shipit that we have open source because that one is designed for continuous deployment of web services. Because the release process in mobile is different and there weren't that many pieces from Shipit that we could reuse, we decided to build it as a separate codebase. I'd definitively recommend companies to build it because then you can build interesting integrations and workflows that otherwise wouldn't be possible. Some CI services are quite flexible and you can leverage that to build your workflows, but that constrains the design of your workflows. We instead imagined how we would like releases to be and build around that.

**Reply by JoeSSS** [reply]
I see, I hoped there is a UI with git coordinator already somewhere :) So we could just use our fastlane tasks on top. We are not yet sure that we want to go this way, so wanted to build a POC first, so trying to find some tooling that can do the POC as quick as possible 😄

**Comment by fabiangrzes** [comment]
Hi 👋, 

 in @Allegro we decided to take a weekly release train approach when submitting an app to the App Store. Release Train is technique for coordinating releases across multiple teams. All releases happen on a fixed and reliable schedule regardless of whether all expected features are ready (the train doesn't wait for you — if you miss it you wait for the next one). Each release is supervised by a tester who performs manually critical test scenarios and a release manager, who is responsible for monitoring mobile app quality and communicating problems to our iOS community on Slack.

 Our release process is divided into two GitHub Workflows. The first one - `Release` starts every Friday at 6:00 AM. This Workflow covers:

 * Creating a release branch. In our case, the release branch specifies app version. By default script bumps minor version e.g. `6.30.0` -> `6.31.0`. For hotfix releases, the release manager creates a release branch manually with bumped patch number and runs the workflow.
 * Creating a JIRA task which allows us to track release progress and sending release notes to Slack release channel.
 * Running tests: Unit, UI and Robo tests, which try to scrap application and reach as many UIViewControllers as possible in order to find easily reproducible crashes. Although such tests aren’t very popular in iOS community, we managed to find some crashes before production stage so we find it useful.
 * After receiving positive tests results, workflow sends release version to TestFlight and DSYMs to crash monitoring tools and commits bumped version to repository. Every build sent to TestFlight bumps build number by one. When build finishes processing, next step submits app for Apple Review. 

 While app awaits review, we perform "stability tests". They cover critical paths, like buying items with a wide range of payment methods. After critical paths tester also performs smoke tests. If everything works fine, app waits for the Apple team’s positive review and is released on Monday. When the review is completed, release manager starts a second workflow `Post Release`. This Workflow covers:

 * Releasing app to App Store
 * Creating tag and pull request to develop branch.
 * Updating `apple-app-site-association` file to server.
 * Creating release version on GitHub.
 * Communicating about the newest release to the release channel on Slack.

 After this stage, release manager starts the process of monitoring app stability using crash monitoring tools.

 For internal builds, we provide for our developers/testers, one of the three options:
 1. Unit Test workflow builds the DEBUG app and uploads it into an artifactory, so everyone could use drag and drop `.app` to simulator for tests. In this case, we use default build and version number.
 2. Build an AdHoc Application using GitHub workflow for testing app on the device. In this case, we use the default build and version number and it's available through GitHub Artifacts.
 3. Build Enterprise Application, which is distributed via Firebase App Distribution, which allows to type custom version and build number, for example, tests backend services. This workflow also allows to modify testers or groups of testers defined in the Firebase Console.
